packages/lesswrong/components/common/PermanentRedirect.tsx

```diff
@@ -1,23 +1,51 @@
 import { registerComponent } from '../../lib/vulcan-lib';
-import React from 'react';
+import React, { useEffect } from 'react';
 import { useServerRequestStatus } from '../../lib/routeUtil'
 import { Redirect } from '../../lib/reactRouterWrapper';
 
+/**
+ * If this component appears in the DOM, this page is a redirect to the given
+ * URL. When rendered server side, it will use the provided HTTP status (default
+ * 301). The URL may be relative or absolute. If there is more than one
+ * PermanentRedirect component in the component tree, behavior is undefined.
+ */
 const PermanentRedirect = ({url, status}: {
   url: string,
   status?: number
 }) => {
   if (!url) throw Error("Permanent Redirect requires a URL")
+  
+  // Handle redirecting in an SSR context. `useServerRequestStatus` will be
+  // null if this is not an SSR.
   const serverRequestStatus = useServerRequestStatus()
   if (serverRequestStatus) {
     serverRequestStatus.status = status || 301
     serverRequestStatus.redirectUrl = url
-  } 
-  return (
-    <Redirect to={url} />
-  );
+  }
+  
+  // Handle redirecting to _absolute_ URLs in a client context. `useEffect`
+  // only runs client side.
+  useEffect(() => {
+    if(urlIsAbsolute(url)) {
+      window.location.replace(url);
+    }
+  }, [url]);
+  
+  // Handle redirecting to _relative_ URLs in a client context. This forwards
+  // to react-router, which only knows how to handle relative URLs not absolute
+  // ones (if you give <Redirect> an absolute URL you might get something like
+  // http://www.lesswrong.com/http://www.lesswrong.com/asdf).
+  if(urlIsAbsolute(url)) {
+    return <></>;
```

Comment:
This is to handle the "what to show while the result of the `window.location.replace` call is loading"?

Reply:
Right; the absolute-URL case rendered client-side is handled inside a `useEffect` so there's a brief time window where this empty-placeholder is in the DOM.

(There's probably a debate to be had about whether the `window.location.replace` should be inside a `useEffect` handler like this, or loose inside the render function. If I understand correctly, the `useEffect` version will have users see a flash of the pre-redirect version, if they got there by client-side navigation, and will work better if the redirect fails.)